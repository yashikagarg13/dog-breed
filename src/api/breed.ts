import { request } from "./index"

const BASE_URL = 'https://dog.ceo/api/'

export function getBreeds() {
  return request(`${BASE_URL}breeds/list/all`) 
}

export  function getImagesByBreed(breed: string) {
  const [mainBreed, subBreed] = breed.split(': ')
  return request(`${BASE_URL}breed/${mainBreed}${subBreed ? '/' : ''}${subBreed ? subBreed : ''}/images/random/5`)
}