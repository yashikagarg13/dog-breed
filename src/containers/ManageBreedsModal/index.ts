export { default } from './ManageBreedsModal'
export type { ManageBreedsModalProps } from './ManageBreedsModal'