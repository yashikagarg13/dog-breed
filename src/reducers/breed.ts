import { createSlice, PayloadAction } from "@reduxjs/toolkit"

interface BreedState {
  loading: boolean;
  fetched: boolean;
  error: string;
  data: string[]
}

const initialState: BreedState = {
  loading: false,
  fetched: false,
  error: '',
  data: []
}

export const breedSlice = createSlice({
  name: 'breed',
  initialState,
  reducers: {
    fetchBreeds: state => {
      state.loading = true
    },
    fetchBreedsSuccess: (state,  action: PayloadAction<string[]>) => {
      state.data = action.payload
      state.loading = false
      state.fetched = true
    },
    fetchBreedsFailure: (state,  action: PayloadAction<string>) => {
      state.data = []
      state.loading = false
      state.error = action.payload
      state.fetched = false
    },
  }
})

export const {
  fetchBreeds,
  fetchBreedsSuccess,
  fetchBreedsFailure,
} = breedSlice.actions


export default breedSlice.reducer