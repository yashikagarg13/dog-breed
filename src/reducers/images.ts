import { createSlice, PayloadAction } from "@reduxjs/toolkit"

export interface Images {
  [key: string]: string[]
}
interface ImagesState {
  loading: boolean;
  fetched: boolean;
  error: string;
  data: Images;
}

const initialState: ImagesState = {
  loading: false,
  fetched: false,
  error: '',
  data: {}
}

export const imagesSlice = createSlice({
  name: 'images',
  initialState,
  reducers: {
    fetchImages: (state,  action: PayloadAction<string[]>) => {
      state.loading = true
    },
    fetchImagesSuccess: (state,  action: PayloadAction<Images>) => {
      state.data = action.payload
      state.loading = false
      state.fetched = true
    },
    fetchImagesFailure: (state,  action: PayloadAction<string>) => {
      state.data = {}
      state.loading = false
      state.error = action.payload
      state.fetched = false
    },
  }
})

export const {
  fetchImages,
  fetchImagesSuccess,
  fetchImagesFailure,
} = imagesSlice.actions


export default imagesSlice.reducer