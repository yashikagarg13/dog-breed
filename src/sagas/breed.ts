import { put, call, takeEvery } from 'redux-saga/effects'
import * as api from '../api/breed'
import {
  fetchBreeds,
  fetchBreedsSuccess,
  fetchBreedsFailure
} from '../reducers/breed'

export function * fetchBreedsRequest(): any {
  try {
    const { message } = yield call(api.getBreeds)
    const breeds = Object.keys(message)
    const subbreeds = breeds
      .filter(breed => breed.length)
      .reduce((combos: string[], breed: string) => {
        return [
          ...combos,
          ...message[breed].map((subbreed: string) => `${breed}: ${subbreed}`)
        ]
      }, [])
    yield put(fetchBreedsSuccess([
      ...breeds,
      ...subbreeds
    ]))
  } catch (e) {
    yield put(fetchBreedsFailure(e.message))
  }
}

export function * watchFetchBreedsRequest () {
  yield takeEvery(
    fetchBreeds,
    fetchBreedsRequest
  )
}
