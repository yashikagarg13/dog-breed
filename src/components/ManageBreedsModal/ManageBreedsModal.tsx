import React, { useEffect, useRef, useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogActions from '@material-ui/core/DialogActions'
import TextField, { TextFieldProps } from '@material-ui/core/TextField'
import Autocomplete from '@material-ui/lab/Autocomplete'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import SearchIcon from '@material-ui/icons/Search'
import BreedList from '../BreedList'
import { ManageBreedsModalProps } from '../../containers/ManageBreedsModal'

const useStyles = makeStyles(theme => ({
  saveButton:  {
    textTransform: 'capitalize',
    background: '#2761F5',
    width: 100,
    fontWeight: 'bold'
  },
  cancelButton: {
    textTransform: 'capitalize',
    background: '#E5E5E5',
    color: '#546079',
    width: 100,
    fontWeight: 'bold'
  },
  error: {
    color: 'red'
  },
  autocompleteOption: {
    textTransform: 'capitalize'
  },
  label: {
    fontWeight: 600,
    margin: 0,
    marginBottom: 5,
    fontSize: 16
  },
  autoCompleteWrapper: {
    marginTop: 30,
    marginBottom: 20,
  },
  dialogTitleRoot: {
    paddingTop: 24,
    paddingBottom: 10
  },
  dialogActionsRoot: {
    padding: '20px 24px'
  }
}))
function ManageBreedsModal(props: ManageBreedsModalProps) {
  const classes = useStyles()
  const { 
    defaultBreeds, 
    open, 
    onClose, 
    onSave,
    breeds,
    error,
    loading,
    fetched,
    fetchBreeds
  } = props
  const [ selectedBreeds, setSelectedBreeds ] = useState<string[]>(defaultBreeds)
  const inputRef = useRef<HTMLDivElement>(null)

  function handleSave() {
    onSave(selectedBreeds)
    onClose()
  }

  function handleChange(e: React.ChangeEvent<{}>, newValue: string[]) {
    if (newValue) {
      setSelectedBreeds(breeds => [...breeds.filter(breed => !newValue.includes(breed)), ...newValue])
      if (inputRef.current) {
        console.log(inputRef.current)
      }
    }
  }

  function handleBreedDelete(selectedBreed: string) {
    const fillteredBreeds = selectedBreeds.filter(breed => breed !== selectedBreed) 
    setSelectedBreeds(fillteredBreeds)
  }

  useEffect(() => {
    if  (!fetched) {
      fetchBreeds()
    }
  }, [fetched, fetchBreeds])

  return (
    <Dialog
      open={open}
      onClose={onClose}
      maxWidth="sm"
      fullWidth
    >
      <DialogTitle classes={{ root: classes.dialogTitleRoot }}>Manage Breeds</DialogTitle>
      <DialogContent>
        <p className={classes.label}>Selected</p>
        <BreedList breeds={selectedBreeds} onDelete={handleBreedDelete} />

        <div className={classes.autoCompleteWrapper}>
          {loading ? (
            <CircularProgress />
          ) : (
            <>
              {error &&  <small className={classes.error}>{error}</small>}
              {breeds?.length && (
                <>
                  <p className={classes.label}>Search for another breed:</p>
                  <Autocomplete
                    multiple
                    id='breed-search'
                    classes={{
                      option: classes.autocompleteOption
                    }}
                    options={breeds}
                    onChange={handleChange}
                    clearOnBlur
                    selectOnFocus
                    value={selectedBreeds}
                    filterSelectedOptions
                    renderInput={(params: TextFieldProps) => 
                      <TextField 
                        {...params} 
                        variant="outlined"
                        size="small"
                        InputProps={{
                          ...params.InputProps,
                          startAdornment: <SearchIcon />
                        }}
                      />}
                  />
                </>
              )}
            </>
          )}
        </div>
      </DialogContent>
      <DialogActions classes={{ root: classes.dialogActionsRoot }}>
        <Button 
          className={classes.cancelButton} 
          variant="contained" 
          color="default" 
          onClick={onClose}>
            Cancel
        </Button>
        <Button 
          className={classes.saveButton} 
          variant="contained" 
          color="primary" 
          onClick={handleSave}>
            Save
        </Button>
      </DialogActions>
    </Dialog>
  )
}

export default ManageBreedsModal