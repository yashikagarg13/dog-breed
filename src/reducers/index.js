import { combineReducers } from 'redux'
import breed from './breed'
import images from './images'

export const getReducersToBeCombined = () => ({
  breed,
  images
})

const reducers = combineReducers(getReducersToBeCombined())
export default reducers
