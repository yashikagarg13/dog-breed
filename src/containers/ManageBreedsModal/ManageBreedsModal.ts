import { connect, ConnectedProps } from 'react-redux'
import { fetchBreeds } from '../../reducers/breed'
import {
  getBreeds,
  getBreedsError,
  isBreedsLoading,
  isBreedsFetched
} from '../../selectors'
import ManageBreedsModal from '../../components/ManageBreedsModal'
import { AppDispatch, RootState } from '../../store'

const mapStateToProps = (state: RootState) => ({
  breeds: getBreeds(state),
  error: getBreedsError(state),
  loading: isBreedsLoading(state),
  fetched: isBreedsFetched(state)
})

const mapDispatchToProps = (dispatch: AppDispatch) => ({
  fetchBreeds: () => dispatch(fetchBreeds())
})

const connector = connect(
  mapStateToProps,
  mapDispatchToProps
)

type ManageBreedsModalPropsRedux = ConnectedProps<typeof connector>

export type ManageBreedsModalProps = ManageBreedsModalPropsRedux & {
  defaultBreeds: string[];
  open: boolean;
  onClose: () => void;
  onSave: (breeds: string[]) => void;
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageBreedsModal) 
