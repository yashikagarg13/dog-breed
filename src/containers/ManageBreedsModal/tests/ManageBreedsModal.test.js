/* global shallowWithStore */
import React from 'react'
import { createMockStore } from 'redux-test-utils'
import toJson from 'enzyme-to-json'

import { fetchBreeds } from '../../../reducers/breed'
import ManageBreedsModal from '../ManageBreedsModal'

const state = {
  breed: {
    loading: false,
    fetched: false,
    error: '',
    data: ['hound']
  }
}

describe('ManageBreedsModal Component', () => {
  let wrapper, store

  beforeEach(() => {
    store = createMockStore(state)
    wrapper = shallowWithStore(<ManageBreedsModal store={store} />)
  })

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should dispatch fetchBreeds', () => {
    wrapper.props().children.props.fetchBreeds()
    expect(store.isActionDispatched(fetchBreeds()))
  })
})
