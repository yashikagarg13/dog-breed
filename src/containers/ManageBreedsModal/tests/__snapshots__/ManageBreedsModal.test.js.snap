// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageBreedsModal Component renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <ManageBreedsModal
    breeds={
      Array [
        "hound",
      ]
    }
    error=""
    fetchBreeds={[Function]}
    fetched={false}
    loading={false}
    store={
      Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
